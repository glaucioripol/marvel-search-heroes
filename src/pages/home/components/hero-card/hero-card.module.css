.card {
  --card-width: 260px;
  --image-width: var(--card-width);
  --image-height: var(--image-width);

  display: flex;
  flex-direction: column;

  width: var(--card-width);

  position: relative;
}

.card .card__figure {
  --border-size: 4px;

  margin: 0;
  padding: 0;
  border-bottom: var(--border-size) solid var(--red-3);
  height: calc(var(--image-height) + var(--border-size));
}

.card .card__content {
  display: flex;
  padding-block: 24px;
  justify-content: space-between;
}

.card .card__title {
  color: var(--grey-2);
  font-size: var(--font-sm);
}
