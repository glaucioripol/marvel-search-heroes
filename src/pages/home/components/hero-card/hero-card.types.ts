import { Hero } from "@/@types/marvel-api-response.types";

export type HeroCardProperties = Readonly<{
  hero: Hero;
  isLiked: boolean;
  handleFavorite: (hero: Hero) => () => void;
}>;
