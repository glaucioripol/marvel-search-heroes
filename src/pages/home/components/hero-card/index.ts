export { HeroCard } from "./hero-card";
