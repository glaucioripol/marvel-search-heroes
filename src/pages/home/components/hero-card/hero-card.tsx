import styles from "./hero-card.module.css";

import { LikeButton } from "@/components";

import { HeroCardProperties } from "./hero-card.types";

export function HeroCard({
  hero,
  isLiked,
  handleFavorite,
}: HeroCardProperties) {
  return (
    <div className={styles.card}>
      <figure className={styles.card__figure} data-testid="card-figure">
        <img
          className={styles.card__image}
          src={`${hero.thumbnail.path}.${hero.thumbnail.extension}`}
          alt={`Imagem do personagem ${hero.name}, não é possível detalhar muito por ser algo dinâmico.`}
          loading="lazy"
          width={260}
          height={260}
        />
      </figure>

      <div className={styles.card__content}>
        <div>
          <h2 className={styles.card__title}>{hero.name}</h2>
        </div>

        <div>
          <LikeButton liked={isLiked} onLikeChange={handleFavorite(hero)} />
        </div>
      </div>
    </div>
  );
}
