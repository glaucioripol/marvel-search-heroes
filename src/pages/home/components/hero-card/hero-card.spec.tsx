import {
  defaultRender,
  fireEvent,
  vi,
  describe,
  expect,
  it,
} from "@/__tests__";

import { HeroCard } from "./hero-card";
import { HeroCardProperties } from "./hero-card.types";

import mockResponse from "@/__fixtures__/characters_default_request.json";

const INDEX_OF_FIRST_HERO = 0;
describe("HeroCard component", () => {
  const mockHero = mockResponse.data.results[INDEX_OF_FIRST_HERO];

  const sutHeroCard = (
    defaultProperties = {
      hero: mockHero,
      isLiked: false,
      handleFavorite: vi.fn(),
    } as HeroCardProperties,
  ) => {
    const renderedComponent = defaultRender(
      <HeroCard {...defaultProperties} />,
    );

    return {
      ...renderedComponent,
      handleFavoriteMock: defaultProperties.handleFavorite,
    };
  };

  it("should render hero name correctly", () => {
    const { getByText } = sutHeroCard();
    const titleElement = getByText(mockHero.name);
    expect(titleElement).toBeInTheDocument();
  });

  it("should render hero image correctly", () => {
    const { getByAltText } = sutHeroCard();
    const imageElement = getByAltText(
      `Imagem do personagem ${mockHero.name}, não é possível detalhar muito por ser algo dinâmico.`,
    ) as HTMLImageElement;
    expect(imageElement).toBeInTheDocument();
    expect(imageElement.src).toBe(
      `${mockHero.thumbnail.path}.${mockHero.thumbnail.extension}`,
    );
  });

  it("should call handleFavorite when like button is clicked", () => {
    const { getByTestId, handleFavoriteMock } = sutHeroCard();
    const likeButton = getByTestId("like-button-default");
    fireEvent.click(likeButton);
    expect(handleFavoriteMock).toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    const { asFragment } = sutHeroCard();
    expect(asFragment()).toMatchSnapshot();
  });
});
