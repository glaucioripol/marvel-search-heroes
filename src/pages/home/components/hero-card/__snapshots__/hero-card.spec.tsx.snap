// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeroCard component > should match snapshot 1`] = `Screen {}`;
